accepted/2020/workloads/workload-manifest.md

```diff
@@ -0,0 +1,430 @@
+
+The .NET tooling will automatically and opportunistically download updated versions of all manifests for the current SDK band and unpack them to `~/.dotnet/sdk-advertising/{sdk-band}/{manifest-id}/`. These user-local updated copies of the manifest are known as *advertising manifests.* The advertising manifests are used when listing workloads that are available to be installed, and for the build tooling to produce warnings that newer versions of installed components are available*.* They are **not** used in pack resolution.
+
+# Format
+
+Manifests are json files. Comments are supported, both `//` and `/* */` styles.
+
+The toplevel is a JSON object, containing the following keys:
+
+| Key | Type | Value | Required |
+|--|--|--|--|
+| `version` | int | The version of the manifest. Must match the version of the NuGet package that contains the manifest. | Yes |
+| `description` | string | Description of the content and/or purpose of the manifest. This is primarily for commenting and/or diagnostic purposes and is not expected to be surfaced in the UX. | No |
+| `workloads` | object | Workload definitions keyed by workload ID. | No |
+| `packs` | object | Pack definitions keyed by pack ID. | No |
+| `data` | object | Allows manifests to include arbitrary data without risk of conflict. | No |
+
+## Workload Definitions
+
+Workload definitions take the following form:
+
+| Key | Type | Value | Required |
+|--|--|--|--|
+| `abstract` | bool | If `true`, this workload can only be extended, and is never exposed directly as an installable workload. Default is `false`. | No |
+| `kind` | string | Either `build` or `dev`. Default is `dev`. | No |
+| `description` | int | User-visible description for the workload. | Yes if dev and non-abstract |
+| `packs` | string array | IDs of the packs that are included in the workload. | No |
+| `extends` | string array | IDs of workloads whose packs should be included in this workload. | No |
+| `platforms` | string array | Limits the workload and workloads that extend it to only be shown and installed on these host platforms. The strings are RIDs. | False |
+
+At least one of `extends` or `packs` is required. If a workload resolves to zero packs, which is possible when some packs are platform-specific, it is implicitly abstract. A workload may transitively include the same pack multiple times or extend the same workload multiple times, and they will be deduplicated. As a consequence, recursive `extends` references are technically permitted but redundant and although they may result in validation warnings they will not result in runtime errors.
+
+The `kind` allows structuring workloads into smaller pieces so that their download and install footprint on CI is smaller. `build` workloads should contain only the packs that are used to build projects. They do not need descriptions as they are not expected to be shown in the UX - they will only be used via a CI-specific UX such as `dotnet workload restore --build-only`.
```

Comment:
As you know, I'm a fan of this. I wonder (in the fullness of time) if we need for of these kinds, to enable more scenarios. One thought is to model the kinds on CLI verbs (`build`, `publish`, `test`, and `run`).

Reply:
Yeah, I'd like to dig into this some more sometime. I think we'll need some real world examples of workloads before we can start trying to divide them into scenario-specific pieces like this. Right now it's just a "do you want it listed or not" flag.